### PR TITLE
[FEATURE][ogr] Add string list field type support

### PR DESCRIPTION
**Pull request by nirvn**
## Description
The title says it all: this PR adds support for OGR's string list field type. It comes in very handy when opening Ordnance Survey's GML datasets.

@nyalldawson , as usual, your review is most, most welcome. Tests preemptively added :wink: 

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nirvn** [comment]
@rouault , if you don't mind, I'd love having your eyes go over this new code too. Merci.

File: src/core/qgsvectorfilewriter.cpp [hunk 1]
```diff
@@ -2253,6 +2263,29 @@ gdal::ogr_feature_unique_ptr QgsVectorFileWriter::createFeature( const QgsFeatur
...
       case QVariant::Invalid:
         break;
+
+      case QVariant::List:
+        // only string list supported at the moment, fall through to default for other types
+        if ( field.subType() == QVariant::String )
+        {
+          QStringList list = attrValue.toStringList();
+          int count = list.count();
+          char **lst = new char *[count];
```
Comment (luipir) [inline src/core/qgsvectorfilewriter.cpp]: who's is in charge to free lst? possible leak?

File: src/core/qgsvectorfilewriter.cpp [hunk 2]
```diff
@@ -2253,6 +2263,29 @@ gdal::ogr_feature_unique_ptr QgsVectorFileWriter::createFeature( const QgsFeatur
...
+          if ( count > 0 )
+          {
+            int pos = 0;
+            for ( QString string : attrValue.toStringList() )
+            {
+              lst[pos] = mCodec->fromUnicode( string ).data();
+              pos++;
+            }
+          }
+          OGR_F_SetFieldStringList( poFeature.get(), ogrField, lst );
```
Comment (luipir) [inline src/core/qgsvectorfilewriter.cpp]: https://www.gdal.org/ogr__api_8h.html#a35dd76723ae1acc91135054e64d7598e => lst should have last element as null ptr as in https://github.com/qgis/QGIS/blob/04c8f5b6f9a4fae67bfc9889ba120f91dcb93193/src/core/qgsvectorfilewriter.cpp#L309

File: src/core/qgsvectorfilewriter.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/qgsvectorfilewriter.cpp]: @luipir , after looking at gdal's source code, we are in charge.

I was seeking enlightenment as to how to free a char **, that'd be my TIL moment ;)

File: src/core/qgsvectorfilewriter.cpp (hunk 2 above)
Comment (nirvn) [reply src/core/qgsvectorfilewriter.cpp]: I shall do.

File: src/core/qgsvectorfilewriter.cpp (hunk 1 above)
Comment (luipir) [reply src/core/qgsvectorfilewriter.cpp]: I'm scared to free after gdal call... no idea in what part or code have to be fee without crash. Here @rouault is necessary

**Comment by nirvn** [comment]
@rouault , is there a reason gdal 2.2.3 would lead to the mass failure of the processing tests under travis? Was the GML driver generated 'fid' field registered as a StringList field in 2.2.X? Also, was there a problem exporting this field type to Shapefile? With my machine, running gdal 2.4, I can't spot any failure.

**Comment by nirvn** [comment]
Suspecting an issue with gdal 2.2.3 (version running on travis), I've made the string list field support conditional to gdal >= 2.4.0 which I've tested locally and confirmed working.

If there are no objections, i'd like to merge this by the end of the day.

**Pull request merged**